substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The stake balance.
+		type Currency: FnInspect<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The staking interface.
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// Something that provides an *always* sorted list of voters.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+
+		/// Something that provides an *always* sorted list of targets by their approval stake.
+		type TargetList: SortedListProvider<
+			Self::AccountId,
+			Score = <Self::Staking as StakingInterface>::Balance,
+		>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A dangling nomination has been successfully dropped.
+		///
+		/// A dangling nomination is a nomination to an unbonded target.
+		DanglingNominationDropped { voter: AccountIdOf<T>, target: AccountIdOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Target is not dangling.
+		///
+		/// A dandling target is a target that is part of the target list but is unbonded.
+		NotDanglingTarget,
+		/// Not a voter/nominator.
+		NotVoter,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		#[cfg(feature = "try-runtime")]
+		fn try_state(_n: BlockNumberFor<T>) -> Result<(), sp_runtime::TryRuntimeError> {
+			Self::do_try_state()
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Removes nomination from a chilled and unbonded target.
+		///
+		/// In the case that an unboded target still has nominations lingering, the approvals stake
+		/// for the "dangling" target needs to remain in the target list. This extrinsic allows
+		/// nominations of dangling targets to be removed.
+		///
+		/// A danling nomination may be removed IFF:
+		///  * The `target` is unbonded and it exists in the target list.
+		///  * The `voter` is nominating `target`.
+		///
+		/// Emits `DanglingNominationDropped`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::drop_dangling_nomination())]
+		pub fn drop_dangling_nomination(
+			origin: OriginFor<T>,
+			voter: AccountIdOf<T>,
+			target: AccountIdOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let _ = ensure_signed(origin)?;
+
+			ensure!(
+				T::Staking::status(&target).is_err() && T::TargetList::contains(&target),
+				Error::<T>::NotDanglingTarget
+			);
+
+			match T::Staking::status(&voter) {
+				Ok(StakerStatus::Nominator(nominations)) => {
+					let count_before = nominations.len();
+
+					let nominations_after =
+						nominations.into_iter().filter(|n| *n != target).collect::<Vec<_>>();
+
+					if nominations_after.len() != count_before {
```

Comment:
Shouldn't this extrinsic essentially be in `pallet-staking`, and then the side effects of it reported here? it seems to achieve the same outcome, but be cleaner. It is a bit icky to think about the fact that this pallet can change the internals of the staking pallet.

Reply:
Yes, I'm convinced that is cleaner and that it's better to keep all changes in staking-related state in the staking pallet -- moved the ext to the staking pallet.